[Offload] Implement the remaining initial Offload API

callumfare opened the pull request:
Implement the complete initial version of the Offload API, to the extent that is usable for simple offloading programs. Tested with a basic SYCL program.

As far as possible, these are simple wrappers over existing functionality in the plugins.

* Allocating and freeing memory (host, device, shared).
* Creating a program 
* Creating a queue (wrapper over asynchronous stream resource)
* Enqueuing memcpy operations
* Enqueuing kernel executions
* Waiting on (optional) output events from the enqueue operations
* Waiting on a queue to finish

Objects created with the API have reference counting semantics to handle their lifetime. They are created with an initial reference count of 1, which can be incremented and decremented with retain and release functions. They are freed when their reference count reaches 0. Platform and device objects are not reference counted, as they are expected to persist as long as the library is in use, and it's not meaningful for users to create or destroy them.

Tests have been added to `offload.unittests`, including device code for testing program and kernel related functionality.

The API should still be considered unstable and it's very likely we will need to change the existing entry points.

callumfare committed:
[Offload] Implement remaining API for basic memory operations and kernel exec

jhuber6 commented on offload/plugins-nextgen/cuda/src/rtl.cpp:
```diff
@@ -1327,6 +1327,34 @@ class CUDAGlobalHandlerTy final : public GenericGlobalHandlerTy {
     return Plugin::success();
   }
+
+  Error getGlobalMetadataFromImage(GenericDeviceTy &Device,
```
Won't we need a version for AMDGPU? There's dynamic_hsa so you should at least be able to ensure it builds.

callumfare replied:
The existing GenericGlobalHandlerTy implementation is enough for it to at least build on AMDGPU, I don't think we need to overload it.

I was a bit unclear in explaining the justification for this change in the call, it's more of a difference between cubin and fatbin that results in the generic ELF lookup failing. I'll write up some detailed documentation when I split this out into a separate PR.

jhuber6 commented on offload/plugins-nextgen/cuda/src/rtl.cpp:
```diff
@@ -1327,6 +1327,34 @@ class CUDAGlobalHandlerTy final : public GenericGlobalHandlerTy {
+                                                                ImageGlobal);
+    }
+
+    CUDADeviceImageTy &CUDAImage = static_cast<CUDADeviceImageTy &>(Image);
```
Isn't this what `getGlobalMetadataFromDevice` is supposed to do?

callumfare replied:
The difference between `getGlobalMetadataFromDevice` and `getGlobalMetadataFromImage` isn't actually that clear to me. Is it that one is intended to look up host-side globals and the other device globals? If that's the case then the proper fix here would probably be to make sure `getGlobalMetadataFromImage` fails gracefully when the binary contains purely device code. Either way I've reverted this change for now.

jhuber6 replied:
It's very distinct, `Image` means "look up up from an ELF I have right here" while `Device` means "Look it up using the vendor API after it's been loaded". The former is way faster.

callumfare replied:
Right, that makes a lot sense, thanks. I'll aim for a more sensible fix in a follow-up PR.

jplehr commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,82 @@
+//===----------------------------------------------------------------------===//
+
+def : Function {
+    let name = "olEnqueueDataWrite";
```
Is there a reason why the name is not explicit of what it does, e.g., `olCopyFromHostToDevice` or should this follow this syntax `ol_copy_from_host_to_device`?
Similar comments apply to the other functions too.

callumfare replied:
There's no particular reason other than the names are based on the matching functions in the plugin interface (substituting read/write for submit/retrieve). The 'enqueue' also makes it clear it's asynchronous rather than a blocking write. This is just personal preference though, I don't have strong opinions, happy to change the names to whatever people think is best.

We discussed the camelCase vs snake_case on the first PR. We slightly leaned towards the former since it seemed to be more common in C APIs (CUDA, OpenCL, etc). Again this is something that we can change though.

jplehr commented on offload/liboffload/include/generated/OffloadEntryPoints.inc:
```diff
@@ -439,3 +439,916 @@ ol_result_t olGetDeviceInfoSizeWithCodeLoc(ol_device_handle_t Device,
+///////////////////////////////////////////////////////////////////////////////
+ol_impl_result_t olMemAlloc_val(ol_device_handle_t Device, ol_alloc_type_t Type,
+                                size_t Size, void **AllocationOut) {
+  if (true /*enableParameterValidation*/) {
```
Question: Is this a flag set when generating the files or picked up from some template?
Wonder if we want a switch to turn this off.

callumfare replied:
This is now controlled with the `OFFLOAD_DISABLE_VALIDATION` env var. It defaults to on, and the unit tests expect it to be on.

jplehr commented on offload/liboffload/include/generated/OffloadEntryPoints.inc:
```diff
@@ -439,3 +439,916 @@ ol_result_t olGetDeviceInfoSizeWithCodeLoc(ol_device_handle_t Device,
+                                               size_t Size,
+                                               void **AllocationOut) {
+  if (offloadConfig().TracingEnabled) {
+    std::cout << "---> olMemAlloc";
```
Do we use `cout` in `offload`?

jhuber6 replied:
I'd prefer not to, they'll probably break if the user updates the global config.

callumfare replied:
This now uses `llvm::errs()`

jplehr reviewed (commented):
Apologies if these questions were already discussed in a meeting or so.

jdoerfert commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +308,298 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+  auto Res = Event->Device->Device.syncEvent(Event->EventInfo);
+  if (Res) {
+    return OL_ERRC_OUT_OF_RESOURCES;
+  }
```
These braces, also other places, should not be here.

callumfare replied:
These should all be fixed now

jdoerfert commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -34,6 +34,69 @@ struct ol_platform_handle_t_ {
   std::vector<ol_device_handle_t_> Devices;
 };
 
+struct ol_queue_handle_t_ {
```
I don't love the '_' suffix idea. It's like there is something missing.

callumfare replied:
I've changed this to `ol_*_queue_impl_t`

jdoerfert commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -34,6 +34,69 @@ struct ol_platform_handle_t_ {
+struct ol_event_handle_t_ {
+  void *EventInfo;
+  ol_queue_handle_t Queue;
+  ol_device_handle_t Device;
```
I'm not sure, but I'm afraid these levels of indirection and duplication might cause a problem at some point. Some of it is also not clear to me. The device associated with a queue cannot change, why do we need a device_handle_t that adds a level of indirection (or is it expected to be flattened later)? For an event, why do we need a Device member that is already in Queue (and I don't think ever used)?

callumfare replied:
I agree that adding a level of indirection here isn't ideal. We should aim to flatten it if possible, or at least keep things to a minimum. Ideally these would just be reference-counting wrappers over the underlying objects. It gets a little complicated if, for example, a user queries which device a queue belongs to - you want to return the `ol_device_handle_t` that they used rather than the underlying GenericDeviceTy, so it helps to keep the handles around.

In this case the `Device` was definitely not needed anyway so I've removed it.

jdoerfert commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +308,300 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+  // TODO: Check error
+  CreatedQueue->Device = Device;
+  CreatedQueue->RefCount = 1;
+  *Queue = CreatedQueue.release();
```
Why do we make unique pointers in the first place? Shouldn't we just create the handle lazily?

callumfare replied:
It's useful when you need the allocation to happen before multiple possible early exits in the functions - means you don't need to remember to free the allocation each time. But I can change this if you'd rather it happens explicitly.

jdoerfert reviewed (commented):
I left a bunch of comments in addition to what we discussed earlier.
There were also some copy-n-paste errors in the td files.

jhuber6 commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +308,298 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+  if (Queue->AsyncInfo->Queue) {
+    auto Err = Queue->Device->Device.synchronize(Queue->AsyncInfo);
+    if (Err) {
+      return OL_ERRC_OUT_OF_RESOURCES;
```
The runtime's handling of errors definitely needs to be updated. Right now we don't have any kind of error codes and fatal error in a few places.

callumfare replied:
I've added `OL_ERRC_UNKNOWN` to cover the cases when the underlying cause isn't actually known. I've added info strings to those errors to give some context though. But yes generally the plugins should be returning more meaningful errors which could be converted directly to Offload errors.

jhuber6 commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -16,8 +16,8 @@ def : Function {
     let details = [];
     let params = [
         Param<"ol_queue_handle_t", "Queue", "handle of the queue", PARAM_IN>,
```
Does the queue know which device it corresponds to? I figured we'd just have a single `olMemcpy` that just copies between two devices. Let the runtime figure out what to do with it.

callumfare replied:
The queue does know which device it belongs to. But I think for a single memcpy to work you'd need to know whether the src and dst pointers are device/shared allocations or host pointers, to know whether to call through to `dataSubmit`, `dataRetrieve`, or `dataExchange` in the plugin interface.

jhuber6 replied:
I think I outlined it like that in my initial draft I just typed up when first thinking about this https://gist.github.com/jhuber6/2117cfd03b7c78d91f5481ac63da682d.

callumfare replied:
That's an interesting idea. A special `host` device could get a bit weird outside of memory operations - could you enqueue kernels and other work on it? It doesn't seem like it would fit in well with the existing host plugin. I think we'd want the host plugin to be a in a usable state wrt memory (targetDataExchange isn't implemented) and kernel execution, to guarantee the host is always present. And also it currently exposes a hard-coded number of devices rather than a single device.

I'd prefer getting something close to what's in this PR in for now, since it reflects the reality of the existing plugins, but with the caveat that we expect it may change soon, and we could discuss an improved design in a future meeting?

callumfare replied:
Sorry I didn't actually see your last comment, I think everything I said above applies though. 

One other issue I can see is who owns host allocations. If you make an allocation on the host plugin, that's not the same as a host allocation made by one of the other plugins (since memory is managed, pinned, etc).

jhuber6 replied:
This is also similar to stuff like CUDA's `cuLaunchHostFunc` which is also a handy way to register callbacks on kernel completion, so I think it's something we should expose.

callumfare replied:
That's all fair, I've added an implementation of a generic async memcpy, along with a function to retrieve the host device. Does it look like what you had in mind?

I've kept the old ones around for comparison, although I've renamed them to memcpy variants for consistency.

I also agree it would be nice to have host task functionality like `cuLaunchHostFunc`.

jhuber6 replied:
Thanks, I think in the future we'll probably be able to get more out of this than just `memcpy`. Likely the whole host plugin should be moved into just 'the current device' and we move the 'plugins' which dlopen some code somewhere else. That should let us do things like enqueue host functions or something, but we can save that for later.

callumfare commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,82 @@
+    let details = [];
+    let params = [
+        Param<"ol_queue_handle_t", "Queue", "handle of the queue", PARAM_IN>,
+        Param<"void*", "SrcPtr", "host pointer to copy from", PARAM_IN>,
```
Yeah this was a weird choice, I've switched them around now

callumfare commented on offload/liboffload/API/Event.td:
```diff
@@ -0,0 +1,41 @@
+
+def : Function {
+    let name = "olReleaseEvent";
+    let desc = "Decrement the reference count of the given event";
```
I've documented this now, and also fixed the release functions that didn't actually delete anything.

callumfare commented on offload/liboffload/include/generated/OffloadAPI.h:
```diff
@@ -530,6 +999,185 @@ typedef struct ol_get_device_info_size_params_t {
+  ol_program_handle_t *pProgram;
+  const char **pKernelName;
+  ol_kernel_handle_t **pKernel;
+} ol_create_kernel_params_t;
```
Yeah these are just to help with printing. They could be moved to the `OffloadPrint.hpp` header to remove some noise from the main API header.

jhuber6 commented:
What is the plan for testing things on the GPU? I'm thinking it would be best to store some C/C++ code that tests the desired functionality and build them separately to a specific name, then mark it as a dependency on the unit test target and just open it as file. I.e.

```
#include <gpuintrin.h>

extern "C" __gpu_kernel void kernel() { some_test(); }
```

Another thing we need to fix in the `libomptarget` runtime is that the images are assumed to live throughout the entire execution, we really need to copy it internally and maintain its lifetime ourself now that it's going to come from external users instead of a constant pointer in an ELF.

jdoerfert commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +315,331 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+
+  auto Err = KernelImpl->init(Device, *Program->Image);
+  if (Err)
+    return {OL_ERRC_UNKNOWN, "Could not initialize the kernel"};
```
I guess later we want to forward `Err` one way or another.

jhuber6 commented on offload/liboffload/include/generated/OffloadEntryPoints.inc:
```diff
@@ -1088,26 +1089,33 @@ olEnqueueKernelLaunch_val(ol_queue_handle_t Queue, ol_kernel_handle_t Kernel,
                       const ol_kernel_launch_size_args_t *LaunchSizeArgs,
                       ol_event_handle_t *EventOut) {
   if (offloadConfig().TracingEnabled) {
     std::cerr << "---> olEnqueueKernelLaunch";
```
```suggestion
    std::cerr << "---> olEnqueueKernelLaunch";
```
No standard streams, use `llvm::errs()` or potentially even a custom logger class.

callumfare commented:
I believe I've addressed most of the comments now so I've closed them, the remaining few I'll look at tomorrow. Let me know if I've missed anything major. Since there are no plugin changes required now I've brought across the changes from the other draft PR and squashed everything together.

The only thing I haven't addressed is the indirection of the handle types (e.g. ol_device_impl_t vs GenericDeviceTy) due to a few issues:

* The `ol_device_impl_t` would have to be 1:1 with `GenericDeviceTy`, and therefore the host device would need to be represented with the actual device from the host plugin, rather than the current placeholder value. There are actually 4 hard-coded devices in the plugin, but even ignoring that it also means that `LIBOMPTARGET_PLUGINS_TO_BUILD` must contain `host` or olMemcpy won't work at all. I don't know if this requirement would break CI or other important configurations.

* Queues and events are represented as void* and require additionally storing the device to be usable. The plugins would ideally have a GenericQueueTy and GenericEventTy to eliminate the indirection required here.

* Reference counts would also have to be done differently, probably as a map of handle pointers to their reference count. There would obviously be a bit of a performance penalty from this but I wouldn't expect it to be unreasonable.

Given the limitations I think it would be better to keep the handle types for now and work towards eliminating them in the future, with the necessary plugin changes.

callumfare commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +315,331 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+}
+
+ol_impl_result_t olCreateKernel_impl(ol_program_handle_t Program,
+                                     const char *KernelName,
```
The interfaces of the implementation functions are identical to the actual entry points, so the string types are kept as `const char*` for consistency. The tablegen backend could convert them to StringRef though if we wanted. Elsewhere we should be consistently using StringRef.

callumfare commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -245,3 +315,331 @@ ol_impl_result_t olGetDeviceInfoSize_impl(ol_device_handle_t Device,
+                                      ol_event_handle_t *EventOut) {
+  if (DstDevice == HostDevice() && SrcDevice == HostDevice()) {
+    // TODO: We could actually handle this with a plain memcpy but we currently
+    // have no way of synchronizing this with the queue
```
Currently the queue is still required, so the memcpy can't be asynchronous. If the queue is made optional though then it would be possible to do the plain memcpy here.

Checking the validity of the queue wrt the devices makes sense though, I'll add that

jhuber6 commented on offload/liboffload/API/Device.td:
```diff
@@ -104,3 +104,15 @@ def : Function {
 }
+
+def : Function {
+  let name = "olGetHostDevice";
```
I still feel like this should be a `kind` when traversing the device list.

jhuber6 replied:
In my mind, the platform is an implementation detail of the device. Something you can query, but the API just iterates over devices.

callumfare replied:
After our discussion in the last call I think we were leaning to keeping the idea of a user-visible platform object rather than just a string or enum property of a device.

There are two ways I can think of doing this:

* Keep the API as is. A user explicitly iterates over platforms, and then fetches devices from those platforms. The platform is a handle that can be queried for information like vendor name etc. Users that don't care about the platforms can just iterate over all of them to get all devices. We would need to add a special host platform to contain the host device.

* Make devices the top-level object that the user interacts with. This essentially means that (as in libomptarget) the user iterates over a flat list of devices. There would still be a platform object that the user can query for from a device (using `olGetDeviceInfo`). This platform handle can then be queried for specific properties. We would need to add a special host platform to contain the host device, but it wouldn't really be user visible unless they go querying for it.

I slightly prefer the first option but I'm happy either way, it's just a small mechanical change in how the API is used to get to the list of available devices.

Is one of these options more preferable? Or is there a better way to go about this?

jhuber6 replied:
I prefer the second, since I think we should really think of this as abstracting over devices, not vendors. However it definitely makes some things more difficulit, because then we need to imagine how to handle things that are incompatible between devices. For example, you can't make an async object for an AMD kernel and then try to schedule an NVIDIA kernel on it.

A possible solution is to allow iterating devices with a filter on it, that's kind of what HSA does. You iterate all the devices and check if it's a GPU or a CPU or whatever.

callumfare replied:
With the current API a user can already filter out specific devices:
```cpp
// For each platform:
size_t NumDevices;
olGetDeviceCount(Platform, &NumDevices);
std::vector<ol_device_handle_t> Devices(NumDevices);
olGetDevices(Platform, NumDevices, Devices.data());
for (auto Device : Devices) {
  ol_device_type_t DeviceType;
  olGetDeviceInfo(Device, OL_DEVICE_INFO_TYPE, sizeof(DeviceType), &DeviceType);

  // Filter out anything that isn't a GPU
  if (DeviceType == OL_DEVICE_TYPE_GPU) {
    FilteredDevices.push_back(Device);
  }
}

```

Similarly platforms (as they're implemented now) can be filtered out based on their `ol_platform_backend_t` property (e.g. only `OL_PLATFORM_BACKEND_AMDGPU`).

All that said we could add a filter parameter to `olGetDevices` like you said, as long as we can add `OL_DEVICE_TYPE_ANY` to allow getting all the devices. It would avoid the need for the above code but wouldn't cover filtering the backends, which we couldn't easily add a filter for if we get rid of `olGetPlatforms`. I'm more in favor of letting the user do the filtering according to their needs with all the information available to them rather than trying to cover all bases with the API, but I'm happy to add the device filter if you'd prefer it.

jhuber6 replied:
HSA does it with callbacks, it's a little clumsy but the most general https://github.com/llvm/llvm-project/blob/main/llvm/tools/llvm-gpu-loader/amdhsa.cpp#L93.

callumfare replied:
That would work - is there any benefit to using a callback that like that versus unconditionally returning the whole list and letting the user do the filtering?

callumfare replied:
I guess actually it's more general than just discovering devices but doing any operation on a number of devices.

jhuber6 replied:
Discovering devices implies initializing the underlying runtime, which has a lot of overhead and resources attached.

callumfare replied:
Right, yeah we talked about avoiding that overhead previously. I think that's another argument to keep platforms as they are just now - a user can filter out platforms with backends they don't want with that callback mechanism (e.g. avoid the platform that reports `OL_PLATFORM_BACKEND_CUDA`). They can then do the same on the devices returned from that platform (e.g. filter out `OL_DEVICE_TYPE_CPU`). That means we can avoid the overhead associated with both steps.

With a flat list of devices we have the issue that we might not know the number of devices and their types until after the runtime is initialized. For example a future OpenCL plugin will need to initialize the OpenCL runtime (in fact likely the OpenCL ICD Loader and multiple runtimes) before knowing about what devices are available to iterate over.

jhuber6 replied:
Hm, it's possible we should have both. Since I feel like the average user will just want to "do this for every device" as the normal path.

callumfare replied:
Yeah we could have a simple `olGetDevices`, and `olGetFilteredDevices` which takes a single callback which can filter on both the platform and device type.

callumfare replied:
Or maybe it needs two callbacks actually, but I think it would be nicer to do with a single API call anyway

jhuber6 replied:
Fine with me either way.

callumfare replied:
I've implemented these changes, see the latest commit.

I'm still on the fence about whether this is better than the previous design. It's more cumbersome to get to the underlying platform handle, and based on the discussion on `olMemFree` we might end up using that more often. We also can't test platforms without any available devices, although that's a pretty minor concern. Another option is to go back to the previous design, but lazily initialize the platforms and devices when they're first used. So returning a platform or device wouldn't initialize the underlying plugin or device, but their first use (e.g. in `olGetDevices(Platform, ...)`, `olCreateQueue(Device, ...)`) would.

jhuber6 replied:
Yeah, I'm not a fan of it. I like how HSA does it with a callback better. OpenMP mostly uses it by more or less getting all devices that match an ELF or LLVM IR file.

callumfare replied:
Would HSA-like `olIteratePlatforms` and `olIterateDevices` be better? The latter taking a platform argument. We could still lazily initialize platforms so just iterating over them doesn't init all the plugins and load dependent runtimes.

jhuber6 replied:
Yeah, probably fine.

callumfare replied:
I've pushed another update based on our last discussion. There's a single HSA style `olDeviceIterate` function and platforms (if needed) can be queried from the device handle. I've also finally moved the host device into the list of regular devices and removed `olGetHostDevice`.

jhuber6 commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,61 @@
+        Param<"ol_event_handle_t*", "EventOut", "optional recorded event for the enqueued operation", PARAM_OUT_OPTIONAL>
+    ];
+    let returns = [
+        Return<"OL_ERRC_INVALID_SIZE", ["`Size == 0`"]>
```
Why is a size of zero an error and not a no-op?

jhuber6 commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,61 @@
+def : Struct {
+    let name = "ol_kernel_launch_size_args_t";
+    let desc = "Size-related arguments for a kernel launch.";
+    let members = [
```
How does this handle dynamic shared memory?

jhuber6 commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,61 @@
+}
+
+def : Function {
+    let name = "olEnqueueKernelLaunch";
```
Not a fan of `EnqueueKernelLaunch` over just `LaunchKernel`.

jhuber6 commented on offload/liboffload/API/Kernel.td:
```diff
@@ -0,0 +1,45 @@
+    let returns = [];
+}
+
+def : Function {
```
What is this needed for? Why do we need a reference count on kernels, it should just be a handle. We could possibly destroy it, but I don't even see a case where a user would want to get rid of a kernel once it's made it. HSA just gives you a symbol lookup + symbol info API. I.e. lookup `foo` and verify `foo` is a kernel.

callumfare replied:
I can remove the kernel handle type as discussed, but one slight disadvantage is that if we make the queue optional for kernel launches then we'd need to additionally pass a device parameter to `olLaunchKernel` so the implementation can create or access the implicit default queue when the queue parameter is null. Not a big deal really but just mentioning it fyi.

jhuber6 commented on offload/liboffload/API/Memory.td:
```diff
@@ -0,0 +1,48 @@
+  let name = "ol_alloc_type_t";
+  let desc = "Represents the type of allocation made with olMemAlloc.";
+  let etors = [
+    Etor<"HOST", "Host allocation">,
```
These names suck, they roughly equate to CUDA's `host`, `managed`, and `device` memory. Honestly we should take this opportunity to change them to something more understandable.

In this context, I believe `host` is 'pinned' memory that always resides on the host `managed` is memory that can migrate in the unified memory context. While device is just memory that only exists on the GPU. HSA has `coarse-grained` and `fine-grained`. coarse-grained being only accessible to one 'agent' (i.e. GPU) while fine-grained is likely pinned. They also have their `svm` API which I believe is closer to `managed`. Naming things is hard unfortunately.

callumfare replied:
I don't disagree but I'm struggling to come up with better names. Maybe we can discuss this in the next call and come to a decision.

jhuber6 replied:
I'd probably just go with `pinned`, `migratable`, and `device` if nothing else. But not going to bikeshed here.

RaviNarayanaswamy replied:
I don't think it should be called pinned.  The memory is allocated on the host, accessible by the device and does not migrate to device.  ie stays on host.  pinned in non offload term could mean pinned on the host and not paged out.
Also the "migratable" memory is  which is accessible by both host and device and the implementation may not migrate the memory but allocate in a location where both host and device can access.

jhuber6 commented on offload/liboffload/API/Memory.td:
```diff
@@ -0,0 +1,48 @@
+  let name = "olMemFree";
+  let desc = "Frees a memory allocation previously made by olMemAlloc.";
+  let params = [
+    Param<"ol_device_handle_t", "Device", "handle of the device to allocate on", PARAM_IN>,
```
This having a device parameter is confounding. If we allocate pinned memory, who owns it? The CPU, the GPU? Both? This is likely something that could be managed inside, but I'm a little weary of putting more pointer trees in runtimes.

callumfare replied:
Yeah that's a good point. We do need some way of dispatching to the correct plugin and like you said it's probably not ideal to track this for all pointers.

The UR equivalent ([`urUSMFree`](https://oneapi-src.github.io/unified-runtime/core/api.html#urusmfree)) takes a context parameter, where the context is a group of one or more devices from the same platform. Host allocations belong to a context rather than any specific device.

I guess the closest we have right now is the platform. The user will always know the platform of the device that did the allocation and can keep a hold of it for the olMemFree call. But by extension that means they could also just track the device itself anyway.

I think in terms of ownership it always makes sense for the GPU to always have ownership because the pinned memory is always managed by a specific device driver and (presumably) won't be compatible between GPUs from different plugins. Likewise shared memory can't migrate between any arbitrary device.

The way `olMemAlloc` is designed, I'm not sure it's meaningful to do, for example, a `OL_ALLOC_TYPE_SHARED` allocation on the special `host` device. I'm not sure how to reconcile these different kinds and the existence of the `host` device as a valid allocation target.

Maybe we need to add the concept of a context, which contains one or more devices and has ownership of the allocations, and are restricted to containing the `host` device and one or more regular devices from a single plugin/platform. The host device could be implicit. Shared memory should always be migratable between the devices, pinned memory should be accessible on every device. Memcpy could take contexts parameters instead of device parameters, and it would still be able to handle copies across different contexts to support the use case of copies between different GPUs etc.

Fundamentally we have to track additional information that underlying APIs don't need to worry about in knowing what platform is managing the memory.

It might be possible to allow contexts to contain any devices but you'd need to be clever with handling pinned memory, and you have problems like device pointers maybe not being unique. I don't know if that would be worth the effort or even solvable.

I'm about to be away for a couple of weeks, but I think this direction is worth discussing, even if we kick the can down the road a bit and leave it for a future PR.

jhuber6 replied:
I'm not an expert, but I think in theory you could have HMM configured for the same memory region between separate GPUs. The HSA API thinks about thins in terms of memory pools and keeps some RB trees to see if a pointer is resident between some allocated ranged IIUC. The `free` function just takes a pointer and then determines the memory pool / device that owns it. They also have the concept of 'setting ownership'. I.e. when you allocate any memory you need to explicitly enable access to the device you want to touch it.

I think in general, we shouldn't force the device here since it'd get really complicated. We could pass the *type* like the CUDA APIs do if we really needed to, but ideally it'd just be a a pointer.

pbalcer replied:
> There's maybe a better way for the plugins to manage it themselves but this seems like an ok first step.

But, without any sort of handle in the free function, you won't know how to dispatch the call to a plugin. In UR, every function has a handle pointer precisely so that it's always possible to route the call to the appropriate underlying adapter/plugin function.

To me this looks like we are trading-off performance and memory (maintaining a DenseMap of **all** allocations won't be cheap and will likely require a mutex) against... convenience of using this API in SYCL/OpenMP implementations?

I don't think that's the right trade-off. If using a device here would be complicated, maybe we should require a platform handle?
The abstraction used across SYCL, OpenCL and Intel Level Zero for this purpose is context.

jhuber6 replied:
A platform handle would be reasonable, but you'd still need to track what kind of allocation it is unless we want to push that onto the user. HSA handles that implicitly, CUDA forces you to remember if it's managed, host, device.

callumfare replied:
With CUDA you can call `cuPointerGetAttributes` to determine whether to call `cuMemFree` or `cuMemFreeHost`. That's how it's implemented in Unified Runtime.

callumfare replied:
I could reinstate the device and kind parameters to remove the map but anything else will require a plugin change:
* The MemoryManager currently belongs to the device, not the platform
* `dataDelete` would need to be changed or an overload added to remove the `TargetAllocTy` argument and let the plugins deduce it if necessary

I'd rather avoid plugin changes in this PR and just expose the existing functionality, to avoid dragging out the review process any longer. I think we're all on the same page that the API is a work in progress so we can make breaking changes to this function after this PR. 

From the Codeplay/Intel side of things, probably our highest priority TODO after this is investigating any gaps between memory capabilities in UR and offload, including whether the equivalent of UR's contexts is needed, which ties into all of this.

jhuber6 commented on offload/liboffload/API/Program.td:
```diff
@@ -0,0 +1,46 @@
+    let name = "olCreateProgram";
+    let desc = "Create a program for the device from the binary image pointed to by `ProgData`.";
+    let details = [
+        "The created program has an initial reference count of 1."
```
What are these reference count things? I don't know why we need them and it massively bloats the API.

callumfare replied:
I've removed all referencing counting now. After some discussion internally we don't think its removal will cause problems for SYCL.

callumfare committed:
Remove reference counting from the API

jhuber6 commented on offload/liboffload/include/generated/OffloadEntryPoints.inc:
```diff
@@ -1088,26 +1089,33 @@ olEnqueueKernelLaunch_val(ol_queue_handle_t Queue, ol_kernel_handle_t Kernel,
       return OL_ERRC_INVALID_NULL_HANDLE;
     }
 
+    if (NULL == ArgumentsData) {
+      return OL_ERRC_INVALID_NULL_POINTER;
+    }
```
C23 has `nullptr` https://en.cppreference.com/w/c/language/nullptr, https://godbolt.org/z/zKs5PTWP6. I think it's better.

callumfare committed:
Rename olEnqueue functions

callumfare committed:
Make queue arguments optional for memcpy and kernel launch

callumfare committed:
Drop device, type parameters from olMemFree

callumfare committed:
Make kernel handles a simple pointer, no create/destroy functions

callumfare committed:
Tidy

callumfare committed:
Add dynamic shared memory arg for olLaunchKernel

jhuber6 commented:
Do we have an interface to iterate symbols? Or do we need to just know the name beforehand.

callumfare commented:
> Do we have an interface to iterate symbols? Or do we need to just know the name beforehand.

There's the generated `OffloadFuncs.inc` file that provides a macro interface to each of the entry points. It's not actually used for anything just now but I figured it would be useful eventually.

callumfare committed:
Implement filtered device/platform discovery, demote platforms

callumfare committed:
Refactor device discovery

callumfare committed:
Rename shared allocations to managed allocations

jhuber6 reviewed (approved):
I think this is fine for now, we can hopefully do some follow-ups in a patch that's less massive. Thanks for sticking with this, at some point I'll need to get some hands-on experience with this so I can make some better judgements. I'm hoping to rewrite the `libc` loader in terms of this API soonish.

callumfare committed:
Fix offload-tblgen tests

callumfare committed:
Remove noisy message from offload test cmake

callumfare committed:
Fix AMDGPU platform discovery

callumfare committed:
Add missing docs

callumfare commented:
@jhuber6 Thanks! I've pushed some small fixes. Would you mind landing this PR when you get a chance? I don't have write access. No new libraries or lit tests this time so hopefully won't be as much trouble as the last one.

jhuber6 commented:
> @jhuber6 Thanks! I've pushed some small fixes. Would you mind landing this PR when you get a chance? I don't have write access. No new libraries or lit tests this time so hopefully won't be as much trouble as the last one.

I'll probably want to try testing it locally first. How do you test it?

callumfare commented:
> > @jhuber6 Thanks! I've pushed some small fixes. Would you mind landing this PR when you get a chance? I don't have write access. No new libraries or lit tests this time so hopefully won't be as much trouble as the last one.
> 
> I'll probably want to try testing it locally first. How do you test it?

From the runtime build dir:
```sh
ninja LibomptUnitTests
./offload/unittests/OffloadAPI/offload.unittests --platform=AMDGPU
```

jhuber6 commented:
Worked when I tested it. But in the future running the `ninja` target needs to run the test with a target like `ninja check-offload-unit` and should auto detect the GPUs on the system.
